Rewrite Syntax for ST4

deathaxe opened the pull request:
This PR completely rewrites LESS.sublime-syntax as extension of ST's default CSS.sublime-syntax using latest syntax engine features.

1. CSS.sublime-syntax is vendored for backward compatibility reasons for all ST builds not shipping https://github.com/sublimehq/Packages/pull/3717

   > **Note**
   >
   > Once a ST build with the PR merged is released, an appropriate release of this package should be added with vendored CSS removed, so it can benefit from future improvements of ST's CSS.

2. Default CSS syntax tests are vendored as syntax_test_css.less to verify all existing CSS expressions are still scoped correctly.

   Only those invalid-syntax tests have been removed, which behave different caused by LESS.

3. Symbol Lists and Symbol Index is refactored to reflect syntax changes and apply some functionality from base CSS.

4. Folding rules for ST's syntax based folding are added

5. Syntax specific settings are added to adjust word separators so `-` is included into identifiers.

#### Benchmark:

| description | before | after | diff
| ---         | -----: | ----: | ---:
| real world CSS file with 10.000 lines of code | 478ms | 30ms | -94%
| syntax_test_less.less from previous revision | 36ms | 8ms | -78%

#### Requirements:

This PR is currently compatible with ST 4126+.

If it is intented to be released with ST4107 [Regexp.sublime-syntax](https://github.com/sublimehq/Packages/blob/master/Regular%20Expressions/RegExp%20(Basic).sublime-syntax) needs to be vendored, too.

deathaxe committed:
Rewrite Syntax for ST4

This commit completely rewrites LESS.sublime-syntax as extension of
ST's default CSS.sublime-syntax using latest syntax engine features.

1. CSS.sublime-syntax is vendored for backward compatibility reasons
   for all ST builds not shipping https://github.com/sublimehq/Packages/pull/3717

   Note: Once a ST build with the PR merged is released, an appropriate
   release of this package should be added with vendored CSS removed,
   so it can benefit from future improvements of ST's CSS.

2. Default CSS syntax tests are vendored as syntax_test_css.less to
   verify all existing CSS expressions are still scoped correctly.

   Those invalid-syntax tests have been removed, which behave different
   caused by LESS.

3. Symbol Lists and Symbol Index is refactored to reflect syntax changes
   and apply some functionality from base CSS.

4. Folding rules for ST's syntax based folding are added

5. Syntax specific settings are added to adjust word separators so `-`
   is included into identifiers.

Benchmark:

- real world CSS file with 10.000 lines of code
  before: 478ms   after: 30ms  diff: -94%

- syntax_test_less.less from previous revision
  before: 36ms    after: 8ms   diff: -78%

deathaxe committed:
Scope CSS Modules from url path

deathaxe commented on Syntaxes/CSS (for LESS).sublime-syntax:
```diff
@@ -0,0 +1,3336 @@
+%YAML 1.2
+---
+# A copy of ST's default CSS to maintain backward compatibility with all builds
```
The idea is to ship this CSS file for ST4xxx - 4148 and create another release branch for ST4149+ with that file removed and LESS directly extending ST's default CSS file.

So older ST builds would benefit from refactored CSS and all users of ST4149+ would use ST's built-in one and thus benefit from future improvements.

Basically the same thing I did for Tailwind CSS recently - except it doesn't ship the CSS for older ST builds.

braver commented:
> This PR is currently compatible with ST 4126+.

Alright, when we're going to ship, we can pre-fix tag stuff.

deathaxe commented:
I am aware of that, but found it of little value to give it a special scope. We'd need to extend the `property-or-selector` context by another set of 1 or 2 branches to scope (maybe parametrised) mixin definitions and mixin calls. It would be possible - is it worth it?

A mixin definition can also look very much like a normal class `#lib.mixin { ... }` So how to distinguish normal classes from mixin definitions?

Without definitions not sure if mixin calls are worth the effort then. The primary use case would be to drive goto definition, symbol list, goto reference.

I'd be otherwise fine to highlight them very much the same like selectors in a first step.

The same is with mixins in values. For simplicity reasons I just scoped them `meta.selector.less` at the moment. The former use of `meta.alias` is wrong, imho.

deathaxe committed:
Use space to align comments

deathaxe commented on Tests/syntax_test_css.less:
```diff
@@ -0,0 +1,3823 @@
+/* SYNTAX TEST "Packages/LESS/Syntaxes/LESS.sublime-syntax" */
```
Yes, it is a copy of core CSS syntax tests to make sure to not break default CSS.

Some tests which CSS would highlight as illegal are left out. Things like nested braces and some illegal nested at-rule tests.

deathaxe commented:
`color:` could be a property or a custom element tag with `:` maybe starting a pseudo class. The current approach to distinguish selectors from properties without maintaining huge and never complete lists of predefined tokens is just to try to scope the thing as `property: value;`. This branch succeeds if a `;` `)` or `}` is found directly after the key-value pair.

https://github.com/danro/LESS-sublime/blob/5dfee6c8eeb0364709c65bdd1a0448f221b8e2c8/Syntaxes/LESS.sublime-syntax#L150-L154

Otherwise an incomplete selector is assumed.

I'd need to look into whether and how it is possible to improve that edge case.

deathaxe committed:
Improve property vs. selector detection

braver commented:
Well color isn’t a valid element and a custom element needs two words and a dash IIRC, but even then to make something work you’d need to maintain a list… and we already agree that that’s not a good idea. 
For me it’s totally acceptable that in some cases the scopes can only be correctly set once the statement is completed. I’ve always found this (selector with pseudo class, or property?) the trickiest bit and never gotten good enough at writing syntaxes to handle it properly. It gets really tricky when a closing semi-colon after a value is not required. That would be an area I would do additional testing.

Other than that, testing with some larger code sets and hunting for edge cases, I defer to you for the scope naming… and it looks really solid already.

deathaxe commented:
HTML also supports "foreign tags", which are any form of valid identifier. A good example are SVG tags. But anything else is fine, too. I guess CSS therefore should be able to highlight them as well. Thus the "custom element needs two words and a dash" rule is unfortunately not sufficient.

I think catching valid CSS at-rules may be something which can easily be added to bailout from property values early, as known at-rules can't be variables and there are not too many.

I aggree that this pseudo-class vs. property issue is very tricky, espececially if no hard coded lists of tokens are to be maintained.

In general CSS's completions try hard to add the semicolon after `:` automatically so it should normally be of no problem to get value completions.

Scope names are mainly just taken from core CSS. I am not too happy with all of them, but I hesitate to change them because it might break many legacy color schemes.

I am looking into adding some more tests about it.

deathaxe committed:
Scope value mixins

deathaxe committed:
Scope property mixins

deathaxe committed:
Scope mixin item access

deathaxe committed:
Vendor ST's RegExp syntax

Maintain compatibility with ST4107-4125.

We could also use builtin regexp, but this would mean to adjust syntax
tests depending on ST builds.

deathaxe commented:
Support for mixins was added and property vs. selector detection slightly improved. Further improvements are probably possible, but that may take some time for investigation and may be done in a future PR as well.

This PR als vendors ST4149's RegExp syntax to maintain compatibility with ST4107-4126 without adjusting syntax tests etc.

I've got an st4149 branch as well which drops both CSS and RegExp.

deathaxe committed:
Fix named arguments

This commit fixes a wrong assumption of `;` denoting mixin definitions.

deathaxe committed:
Scope named arguments variable.parameter

Exclude them from normal goto definition / goto reference functionality.

deathaxe committed:
Re-scope parameter separators

deathaxe committed:
Refactor mixin call arguments lists

This commit...
1. introduces a `less-values` context to avoid messing with
   `less-group-body`.
2. Fixes detection of mixin calls of detached rulesets.

deathaxe committed:
Refactor parametrized mixins

This commit ...

1. scopes mixin parameters of parametrized mixins `variable.parameter`
2. scopes parameter lists `meta.selector.parameters meta.group`
3. adjusts selector's symbol list definition

This way scoping/highlighting is adjusted to how parameter/arguments
lists of function definitions vs. function calls are scoped.

deathaxe committed:
Refactor Indexed Symbol/Reference Lists

1. more specific selectors with regards to exclude `punctuation`
2. add mixins to reference index to enable Goto Reference feature
3. add parametrized mixin parameters to symbol index to enable
   Goto Definition feature

deathaxe committed:
Unify mixin arguments

This commit uses same contexts for both property-level and value-level mixins.

deathaxe committed:
Fix property mixins

Both property-level mixins and values are terminated by the same tokens.
By also including comments, another edge case in detached rulesets is fixed
where mixins were scoped `meta.selector.css` before.

deathaxe committed:
Move less-values context

Seems better placed next to less-expressions

braver commented on Syntaxes/LESS.sublime-syntax:
```diff
@@ -1,2691 +1,1042 @@
-  - less
-  - css.less
+# https://lesscss.org
+name: LESS
```
Actually the name used on https://lesscss.org is Less (lowercase "ess").

braver commented on Syntaxes/LESS.sublime-syntax:
```diff
@@ -1,2691 +1,1042 @@
+###[ CSS MODULES / PSEUDO CLASSES ]############################################
+
+  css-modules-pseudo-class:
+    # https://github.com/css-modules/icss#specification
```
I thought css-modules is a dead project. I kept it around for compatibility, but is it worth it?

deathaxe replied:
Idk. Re-added it as it was already there. We can simply drop it, if it is of no use anymore.

deathaxe replied:
It doesn't hurt though. Maybe some legacy code needs it, still.

braver replied:
Sure, it doesn't hurt and it's already implemented. I have no idea about the adoption rate of it... So, let's keep it. 👌🏻

deathaxe committed:
Change Name Casing

deathaxe committed:
Fix typo

deathaxe committed:
Change Name Casing of Embeddings

deathaxe committed:
Add bailout from incomplete mixin brackets

Syntax after incomplete mixin brackets may break until end of document without
a 2nd-level bailout.

braver commented:
Alright, looking good… let’s do it like this:

- You move your fork to the Sublime Text org
- Set up the correct prefixed tags for backwards compatibility 
- We can keep this PR open so I have another chance to look at it and maybe give some feedback next week
- Set up a PR on package control to replace the Less entry with the fork

deathaxe commented:
All prepared. Repo moved, you added as maintainer, tags added retarget PR filed.

braver commented on Tests/syntax_test_less.less:
```diff
@@ -1,1925 +1,2756 @@
+//^^^^^^^ meta.declaration.identifier.less variable.other.less
+//       ^ meta.declaration.less punctuation.separator.key-value.css
+//        ^^^^^^^^^^ meta.declaration.value.less
+//         ^^^^^^^^^ meta.mixin.less entity.other.attribute-name.class.css
```
Honestly, Less breaks my brain a little bit. 

> By combining namespacing with the lookup [] syntax, you can turn your rulesets / mixins into maps.

Wut. Why don't they introduce data types as separate things, everything it constantly turning into something else by slight syntax or context changes. Maybe we shouldn't try to attach too much meaning to anything as long as it highlights somewhat correctly.

Anyway, in this particular case, besides any selector of course potentially being a mixing, and besides any value of a variable being nothing more than a string that at some later point might evaluate into a property, a value, a map or a selector, or something else... I don't understand why `.selector` should have the `meta.mixin.less` scope?

braver replied:
> But that's the cost of adding (too much) meaning

Totally agree. I believe Less is intentionally ambiguous in many situations; it all makes sense when you parse and render the entire thing to CSS, but that's not what a syntax is for. 

> Thus can't use them to distinguish selectors and mixins

Right, well there goes that idea. 

> treates every selector followed by { as normal selector (meta.selector) and those followed by ; as mixins (meta.mixin).

That's reasonable, but also potentially incorrect. In `@variable: .selector();` the `.selector()` is definitely a mixin and not a selector, but in `@variable: .selector;` the `.selector` is _probably_  just a selector and because it's a selector it's _potentially_ a mixin. Honestly, maybe we're better off not even trying to implement the entire selector/mixin thing. The language purposely foregoes the concept of explicit declaration of mixins, so you can never do "goto definition", "find references" anyway just based on the syntax. I doubt even an LSP can make heads or tails of it without knowing exactly how and in which order `@import`s are used (ie. what the entry point for compilation is). 
So let's go with the simplest thing that still makes some sense. I don't have a real objection against your current implementation, so you can leave it as is. And do whatever you like with this feedback of course.

deathaxe replied:
We can keep it in mind as possible enhancement and revisit later.

michaelblyons commented:
Package Control PRs at
- https://github.com/wbond/package_control_channel/pull/8749
- https://github.com/wbond/package_control_channel/pull/8751
